lib/Frontend/ModuleInterfaceLoader.cpp

```diff
@@ -1406,20 +1406,25 @@ ModuleInterfaceCheckerImpl::getCompiledModuleCandidatesForInterface(StringRef mo
     llvm::sys::path::replace_extension(modulePath, newExt);
   }
 
-  ModuleInterfaceLoaderImpl Impl(Ctx, modulePath, interfacePath, moduleName,
-                                 CacheDir, PrebuiltCacheDir, BackupInterfaceDir,
-                                 SourceLoc(), Opts,
-                                 RequiresOSSAModules,
-                                 RequireNCGenerics,
-                                 nullptr,
-                                 ModuleLoadingMode::PreferSerialized);
+  ModuleInterfaceLoaderImpl Impl(
+      Ctx, modulePath, interfacePath, moduleName, CacheDir, PrebuiltCacheDir,
+      BackupInterfaceDir, SourceLoc(), Opts, RequiresOSSAModules,
+      RequireNCGenerics, nullptr, ModuleLoadingMode::PreferSerialized);
   std::vector<std::string> results;
   auto pair = Impl.getCompiledModuleCandidates();
-  // Add compiled module candidates only when they are non-empty.
-  if (!pair.first.empty())
-    results.push_back(pair.first);
-  if (!pair.second.empty())
+  auto validateModule = [&](StringRef modulePath) {
+    SmallVector<FileDependency, 16> deps;
+    std::unique_ptr<llvm::MemoryBuffer> moduleBuffer;
+    return Impl.upToDateChecker.swiftModuleIsUpToDate(
+        modulePath, Impl.rebuildInfo, deps, moduleBuffer);
+  };
+
+  // Add compiled module candidates only when they are non-empty and up-to-date.
+  // Since we are scanning interface, prefer the one in the cache.
+  if (!pair.second.empty() && validateModule(pair.second))
```

Comment:
We should really make this a struct so the modules are named... If you don't want to change that in this PR you could use `std::tie(adjacentMod, prebuiltMod) = ` to destructure the pair.